### ISSUE TITLE
`clippy::redundant_closure`: false positive when calling `vec![]`

### ISSUE THREAD

**Issue by remi-dupre** [rust-lang/rust-clippy#7435]
Hi!

Here is a minimal example to highlight the issue:

```rust
pub fn get_result() -> Result<Vec<i32>, String> {
    Ok(vec![0, 2, 4])
}

pub fn weird() {
    // This line triggers clippy::redundant_closure:
    // warning: redundant closure
    // --> src/lib.rs:6:42
    //  |
    //  |     let _x = get_result().unwrap_or_else(|_| vec![]);
    //  |                                          ^^^^^^^^^^ help: replace the closure with `Vec::new`: `std::vec::Vec::new`
    //  |
    //  = note: `#[warn(clippy::redundant_closure)]` on by default
    //  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
    let _x = get_result().unwrap_or_else(|_| vec![]);

    // This line doesn't trigger any warning
    let _y = get_result().unwrap_or_else(|_| Vec::new());
    let _z = get_result().unwrap_or_else(|_| vec![0]);
}
```

`clippy` suggests to update the code with something invalid as `Vec::new` is not equivalent to `|_| vec![]` which takes a parameter. 

I think this may be related to [#3071](https://github.com/rust-lang/rust-clippy/issues/3071) but still seems to be a different issue since the `vec![]` macro call appears to be what causes it (there is no problem using `Vec::new()`).

### Meta

- `cargo clippy -V`: clippy 0.1.53 (53cb7b0 2021-06-17)
- `rustc -Vv`:
  ```
  rustc 1.53.0 (53cb7b09b 2021-06-17)
  binary: rustc
  commit-hash: 53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b
  commit-date: 2021-06-17
  host: x86_64-unknown-linux-gnu
  release: 1.53.0
  LLVM version: 12.0.1
  ```

**Comment by remi-dupre** [comment]
I've just noticed that the issue doesn't show up when using the nightly: `cargo +nightly clippy`. Does this mean that the issue is already fixed and will be available in next release?

**Comment by giraffate** [comment]
Yes, it will be fixed at 1.54.0. FYI changelog is here: <https://github.com/rust-lang/rust-clippy/blob/59a164e86c41b4ccfa1e41ec60d7bb8a0f2c87d2/CHANGELOG.md#false-positive-fixes>.

I'm closing this. Anyway, thanks for the report!
